04_Scheduling/01_README.md

```diff
@@ -0,0 +1,16 @@
+# Scheduling and Tasks
+
+Up until this point, we've be running a single stream of code. Getting the cpu and other hardware setup how we want. However we very quickly run into a few issues: what if we want to wait on some hardware, but with a timeout? What if we want to run multiple streams of code at the same time? Multiple programs with multiple threads?
+This is where a scheduler comes in. It's similar to a hardware multiplexer (sometimes called a mux), in that it takes multiple inputs (our code) and allows them to share a single output (our cpu, which can only run one instruction stream). 
+
+The scheduler works by interrupting the currently running program, saving it's state and then loading the state of a new program. This is called a context switch, and if often quite an expensive and slow operation to perform, as we'll see later. As you might have guessed, the scheduler often operates within an interrupt handler, and if written correctly, the running program will never know it has paused at all. If these context switches are performed at the right speed, the user will never notice programs briefly pausing and unpausing, and it will appear to the user as if all programs are running at the same time.
+
+Of course unless you support multiple cores, we know this is far from the truth. It's all just a clever trick.
```

Comment:
I know I wrote this, but now I'm thinking that perhaps a good analogy could be how videos work: It's just a sequence of still images played rapidly enough that we perceive it as smooth motion.